File: 791. Custom Sort String/CustomSortString.py
'''
https://leetcode.com/problems/custom-sort-string/
'''


class Solution:
    # O(n) time, depends on our string s
    # O(1) space, we are just storing alphabets in a map so 26 len is constant
    # Approach: Hashmap, sorting
    def customSortString(self, order: str, s: str) -> str:
        n = len(order)
        order_map = {}
        ch_count = {}
        ans = ""
        
        for ch in s:
            ch_count[ch] = ch_count.get(ch, 0) + 1
            if ch in order:
                order_map[ch] = order.index(ch) + 1
            else:
                order_map[ch] = n + 1
                
        sorted_str = sorted(order_map.items(), key=lambda x:x[1])
        
        for ch, weight in sorted_str:
            ans += (ch * ch_count[ch])
            
        return ans

File: 791. Custom Sort String/custom_sort_string.rs
use std::collections::HashMap;


impl Solution {
    
    // O(n) time,
    // O(n) space,
    // Approach: hashmap, counting
    pub fn custom_sort_string(order: String, s: String) -> String {
        let mut count = HashMap::new();
        for ch in s.chars() {
            *count.entry(ch).or_insert(0) += 1;
        }
        
        let mut ans: String = Default::default();
        for ch in order.chars() {
            if let Some(ch_count) = count.get(&ch) {
                ans.push_str(&ch.to_string().repeat(*ch_count as usize));
                count.remove(&ch);
            }
        }
        
        for (ch, ch_count) in count {
            ans.push_str(&ch.to_string().repeat(ch_count));
        }
        
        ans
    }
}